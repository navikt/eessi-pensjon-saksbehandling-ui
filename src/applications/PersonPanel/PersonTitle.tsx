import kvinne from 'src/assets/icons/icon-kvinne.png'
import mann from 'src/assets/icons/icon-mann.png'
import ukjent from 'src/assets/icons/icon-ukjent.png'
import classNames from 'classnames'
import { PersonPDL } from 'src/declarations/person'
import _ from 'lodash'
import moment from 'moment'
import {Heading, HStack, Link} from '@navikt/ds-react'
import {getFnr, getNPID} from 'src/applications/BUC/components/BUCUtils/BUCUtils'
import LoadingImage from "src/components/Loading/LoadingImage";
import {copyToClipboard} from "src/actions/app";
import {useDispatch} from "react-redux";
import {useState} from "react";
import {useTranslation} from "react-i18next";
import {FilesFillIcon, FilesIcon} from "@navikt/aksel-icons";
import styles from "src/assets/css/common.module.css";
import titleStyles from "./PersonTitle.module.css"

export interface PersonTitleProps {
  gettingPersonInfo: boolean
  person?: PersonPDL
}

const PersonTitle: React.FC<PersonTitleProps> = ({
  gettingPersonInfo, person
}: PersonTitleProps): JSX.Element => {
  const [hover, isHover] = useState<boolean>(false);
  const dispatch = useDispatch()
  const { t } = useTranslation()

  let birthDate: Date | undefined
  let deathDate: Date | undefined

  let age: number = 0

  if (!person || gettingPersonInfo) {
    return (
      <LoadingImage />
    )
  }

  if (_.get(person, 'foedselsdato.foedselsdato')) {
    birthDate = moment(person.foedselsdato.foedselsdato)?.toDate()
  }
  if (_.get(person, 'doedsfall.doedsdato')) {
    deathDate = moment(person.doedsfall?.doedsdato)?.toDate()
  }

  const endDate: Date = deathDate || new Date()
  if (birthDate && endDate) {
    age = moment(endDate).diff(moment(birthDate), 'years')
  }

  let kind: string = 'nav-unknown-icon'
  let src = ukjent

  if (person.kjoenn?.kjoenn === 'KVINNE') {
    kind = 'nav-woman-icon'
    src = kvinne
  } else if (person.kjoenn?.kjoenn === 'MANN') {
    kind = 'nav-man-icon'
    src = mann
  }

  const pid : string | undefined = getFnr(person) ? getFnr(person) : getNPID(person)

  return (
    <HStack gap="4" align="center">
      <img
        alt={kind}
        className={classNames({ [titleStyles.dead]: !_.isNil(deathDate) })}
        data-testid='w-persontitle--img'
        src={src}
        width={40}
      />
      <Heading size='medium'>
        {person.navn?.sammensattNavn} ({age}) - {pid}
        <Link
          onMouseEnter={() => isHover(true)}
          onMouseLeave={() => isHover(false)}
          title={t('ui:person-kopier-pid')} onClick={(e: any) => {
          e.preventDefault()
          e.stopPropagation()
          pid && dispatch(copyToClipboard(pid))
        }}
        >
          {hover ?
            <FilesFillIcon className={styles.copyFilledWithMargin} fontSize="1.5rem"/> :
            <FilesIcon className={styles.copyWithMargin} fontSize="1.5rem"/>
          }
        </Link>
      </Heading>
    </HStack>
  )
}

export default PersonTitle
