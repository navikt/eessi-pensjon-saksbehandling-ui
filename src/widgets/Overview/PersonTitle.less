.w-overview-personPanel__icon.dead {
  filter: grayscale(100%)
}
