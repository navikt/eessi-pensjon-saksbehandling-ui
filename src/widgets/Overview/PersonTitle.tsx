import kvinne from 'assets/icons/icon-kvinne.png'
import mann from 'assets/icons/icon-mann.png'
import ukjent from 'assets/icons/icon-ukjent.png'
import classNames from 'classnames'
import { HorizontalSeparatorDiv } from 'components/StyledComponents'
import { Person } from 'declarations/types'
import { PersonPropType } from 'declarations/types.pt'
import _ from 'lodash'
import moment, { Moment } from 'moment'
import { Systemtittel } from 'nav-frontend-typografi'
import PT from 'prop-types'
import React from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'
import PersonLoading from 'widgets/Overview/PersonLoading'

export interface PersonTitleProps {
  gettingPersonInfo: boolean
  person?: Person
  personAvdod: any
}

const Title = styled.div`
  display: flex;
  align-items: center;
  .dead {
    filter: grayscale(100%)
  }
`
const PersonTitle: React.FC<PersonTitleProps> = ({
  gettingPersonInfo, person, personAvdod
}: PersonTitleProps): JSX.Element => {
  let birthDate: Date | Moment | undefined
  let deathDate: Date | Moment | undefined
  const { t } = useTranslation()
  if (!person || gettingPersonInfo) {
    return (
      <PersonLoading />
    )
  }

  if (_.get(person, 'foedselsdato.foedselsdato')) {
    birthDate = moment(person.foedselsdato.foedselsdato)
    if (birthDate) {
      birthDate = birthDate.toDate()
    }
  }
  if (_.get(person, 'doedsdato.doedsdato')) {
    deathDate = moment(person.doedsdato.doedsdato)
    if (deathDate) {
      deathDate = deathDate.toDate()
    }
  }

  const age: number = (deathDate ? (deathDate as Date).getFullYear() : new Date().getFullYear()) - (birthDate as Date).getFullYear()

  let kind: string = 'nav-unknown-icon'
  let src = ukjent
  if (person.kjoenn.kjoenn.value === 'K') {
    kind = 'nav-woman-icon'
    src = kvinne
  } else if (person.kjoenn.kjoenn.value === 'M') {
    kind = 'nav-man-icon'
    src = mann
  }

  return (
    <Title>
      <img
        alt={kind}
        width={40}
        src={src}
        className={classNames({ dead: !_.isNil(deathDate) })}
      />
      <HorizontalSeparatorDiv />
      <Systemtittel>
        {person.personnavn.sammensattNavn} ({age}) - {person.aktoer.ident.ident}
        {personAvdod !== undefined ? ' - ' + t('ui:deceased') + ': ' + personAvdod.length : ''}
      </Systemtittel>
    </Title>
  )
}

PersonTitle.propTypes = {
  gettingPersonInfo: PT.bool.isRequired,
  person: PersonPropType
}

export default PersonTitle
