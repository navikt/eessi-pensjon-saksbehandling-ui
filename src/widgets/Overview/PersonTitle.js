import React from 'react'
import PT from 'prop-types'
import _ from 'lodash'
import classNames from 'classnames'
import { Icons, Nav, WaitingPanel } from 'eessi-pensjon-ui'
import './PersonTitle.css'

const PersonTitle = ({ gettingPersonInfo, person, t }) => {
  let birthDate
  let deathDate

  if (!person) {
    return null
  }

  if (gettingPersonInfo) {
    return <WaitingPanel className='w-overview-personPanel__waiting' message={t('ui:loading')} />
  }

  if (person && person.foedselsdato && person.foedselsdato.foedselsdato) {
    birthDate = new Date(Date.parse(person.foedselsdato.foedselsdato))
  }
  if (person && person.doedsdato && person.doedsdato.doedsdato) {
    deathDate = new Date(Date.parse(person.doedsdato.doedsdato))
  }

  const age = (deathDate ? deathDate.getFullYear() : new Date().getFullYear()) - birthDate.getFullYear()

  let kind = 'nav-unknown-icon'
  if (person.kjoenn.kjoenn.value === 'K') {
    kind = 'nav-woman-icon'
  } else if (person.kjoenn.kjoenn.value === 'M') {
    kind = 'nav-man-icon'
  }

  return (
    <div className='w-overview-personPanel__title'>
      <Icons size={40} kind={kind} className={classNames('w-overview-personPanel__icon', { dead: !_.isNil(deathDate) })} />
      <Nav.Systemtittel className='ml-2'>
        {person.personnavn.sammensattNavn} ({age}) - {person.aktoer.ident.ident}
      </Nav.Systemtittel>
    </div>
  )
}

PersonTitle.propTypes = {
  aktoerId: PT.string,
  gettingPersonInfo: PT.bool,
  person: PT.object,
  t: PT.func.isRequired
}

export default PersonTitle
