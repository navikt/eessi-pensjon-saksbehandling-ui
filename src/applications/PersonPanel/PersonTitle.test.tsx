import { render } from '@testing-library/react'
import PersonTitle, { PersonTitleProps, Title } from './PersonTitle'

import mockPerson from 'mocks/person/personPdl'

describe('applications/PersonPanel/PersonTitle', () => {
  let wrapper: any

  const initialMockProps: PersonTitleProps = {
    gettingPersonInfo: false,
    person: mockPerson
  }

  beforeEach(() => {
    wrapper = render(<PersonTitle {...initialMockProps} />)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('Render: match snapshot', () => {
    const { container } = render(<PersonTitle {...initialMockProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('Render: waiting spinner if no person is given', () => {
    wrapper = render(<PersonTitle {...initialMockProps} person={undefined} />)
    expect(wrapper.exists('PersonLoading')).toBeTruthy()
  })

  it('Render: has proper HTML structure', () => {
    expect(wrapper.exists(Title)).toBeTruthy()
    expect(wrapper.find(Title).find('img').props().alt).toEqual('nav-man-icon')
    expect(wrapper.find(Title).find('h2').render().text()).toEqual('LEALAUS SAKS (41) - personFnr')
  })

  it('Render: different person icons', () => {
    mockPerson.kjoenn.kjoenn = 'MANN'
    wrapper.setProps({ person: mockPerson })
    expect(wrapper.find('[data-testid=\'w-persontitle--img\']').props().alt).toEqual('nav-man-icon')

    mockPerson.kjoenn.kjoenn = 'UKJENT'
    wrapper.setProps({ person: mockPerson })
    expect(wrapper.find('[data-testid=\'w-persontitle--img\']').props().alt).toEqual('nav-unknown-icon')

    mockPerson.kjoenn.kjoenn = 'KVINNE'
    wrapper.setProps({ person: mockPerson })
    expect(wrapper.find('[data-testid=\'w-persontitle--img\']').props().alt).toEqual('nav-woman-icon')
  })
})
