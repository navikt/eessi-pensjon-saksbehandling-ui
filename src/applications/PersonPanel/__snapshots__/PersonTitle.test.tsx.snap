// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applications/PersonPanel/PersonTitle Render: match snapshot 1`] = `
.c3 {
  display: inline-block;
  margin-left: 1rem;
}

ui-fss.g18[id="c2"] {
  content: "c3,";
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 .dead {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
}

ui-fss.g44[id="c0"] {
  content: "c1,";
}

<div
  class="c0 c1"
>
  <img
    alt="nav-man-icon"
    class=""
    data-testid="w-persontitle--img"
    src="icon-mann.png"
    width="40"
  />
  <div
    class="c2 c3"
  />
  <h1
    class="navds-heading navds-heading--medium"
  >
    LEALAUS SAKS
     (
    42
    ) - 
    personFnr
  </h1>
</div>
`;
